Allow custom ACME server

tashian opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Coding guidelines: -->

------------------------------

 - [x] I have read the [contributing](https://github.com/linuxserver/docker-swag/blob/master/.github/CONTRIBUTING.md) guideline and understand that I have made the correct modifications

------------------------------

## Description:
<!--- Describe your changes in detail -->
Support for a custom ACME server and CA bundle.
I've added `ACMESERVER` and `ACMECABUNDLE`  variables to support this.
The root CA bundle in `ACMECABUNDLE` is expected to be a base64 encoded PEM file.
The config init script decodes and saves the CA bundle to `/config/cabundle.pem`.
The variable `REQUESTS_CA_BUNDLE` is exported for all calls to `certbot`, so that `certbot` can establish CA trust.

## Benefits of this PR and context:
<!--- Please explain why we should accept this PR. If this fixes an outstanding bug, please reference the issue # -->
Internal ACME servers have become more popular and should be supported.
Closes #186.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Not yet tested.

## Source / References:
<!--- Please include any forum posts/github links relevant to the PR -->
There's no formal documentation that I know of for the `REQUESTS_CA_BUNDLE` env variable passed into `certbot`, but [this blog post](https://smallstep.com/blog/private-acme-server/) shows it in action.

tashian commented:
Just wanted to note that I don't love having to send a base64-encoded root CA PEM into the container via an env variable.
Maybe there is another approach to getting this file into the container?

tashian commented:
Bump

benoitj commented:
@tashian on alpine you can pkg add ca-certificates, store a ca cert in /usr/local/share/ca-certificates directory and run update-ca-certificates. this seems to make the ca cert available for all tools.

tested this on a alpine docker container, makes curl work on my custome ACME server

brycied00d commented:
Bumping this to demonstrate my own interest in this feature.

tashian commented:
bump

coreyramirezgomez commented on root/etc/cont-init.d/50-config:
```diff
@@ -21,7 +23,7 @@ if [ -n "${TEST_RUN}" ]; then
 
 # Sanitize variables
-SANED_VARS=( DNSPLUGIN EMAIL EXTRA_DOMAINS ONLY_SUBDOMAINS STAGING SUBDOMAINS URL VALIDATION CERTPROVIDER )
+SANED_VARS=( DNSPLUGIN EMAIL EXTRA_DOMAINS ONLY_SUBDOMAINS STAGING SUBDOMAINS URL VALIDATION CERTPROVIDER ACMESERVER ACMECABUNDLE )
```
ACMECABUNDLE gets mangled by the sanitizer here. I have removed it from this list and everything works fine after this.

coreyramirezgomez commented:
I am very interested in seeing this added as a feature. I have an internal CA (via step-ca + ACME provisioner) I have tested against for which I can provide results that this indeed does work (minus one change which I have already highlighted).

aptalca commented:
We've been working on a big rebase of our nginx baseimage and new features were put on hold. We're very close to completing that so we'll take a look at this soon.

aptalca commented:
@tashian
our big nginx rebase just got merged to SWAG so we are now reviewing PRs for new features again.

If you rebase this, we'll review and merge soon

Thanks

tashian commented:
Hi @aptalca, thanks for the update. I've rebased this PR.

tashian commented:
@coreyramirezgomez thanks for testing on your setup! I committed your suggestion

tashian commented:
Ok, I've addressed these issues, thanks for the feedback.

I also updated the logic to allow a file-mounted CA bundle. I think this is more straightforward than providing a base64-encoding PEM file in an env variable, but I know this option won't be available in some container environments.

aptalca commented on root/etc/cont-init.d/50-certbot:
```diff
@@ -53,23 +55,40 @@ if [ -z "$VALIDATION" ]; then
-    ACMESERVER="https://acme-staging-v02.api.letsencrypt.org/directory"
+# Choose the relevant CA server
+if [ -n "$ACMEURL" ]; then
+	if [ -z "$EMAIL" ]; then
```
`CERTPROVIDER` value should be used to make this determination. A user could have set `ACMEURL`, but if `CERTPROVIDER` is not set to `custom`, this part should not halt the process.

aptalca commented on root/etc/cont-init.d/50-certbot:
```diff
@@ -229,7 +248,7 @@ else
 
 # Check if the cert is using the old LE root cert, revoke and regen if necessary
-if [ -f "/config/keys/letsencrypt/chain.pem" ] && ([ "${CERTPROVIDER}" == "letsencrypt" ] || [ "${CERTPROVIDER}" == "" ]) && [ "${STAGING}" != "true" ] && ! openssl x509 -in /config/keys/letsencrypt/chain.pem -noout -issuer | grep -q "ISRG Root X"; then
+if [ -f "/config/keys/letsencrypt/chain.pem" ] && ([ "${CERTPROVIDER}" = "letsencrypt" ] || ([ "${CERTPROVIDER}" = "" ] && [ -z "$ACMECABUNDLE" ])) && [ "${STAGING}" != "true" ] && ! openssl x509 -in /config/keys/letsencrypt/chain.pem -noout -issuer | grep -q "ISRG Root X"; then
```
This new logic is incorrect. In fact, this logic does not need to be modified at all, as it is just a backwards compatibility shim for a breaking change from well over a year ago.

The part that needs to be modified is the following, to make sure that the changes to the custom ACME related vars result in revoking the old cert and forcing revalidation:
https://github.com/linuxserver/docker-swag/blob/fe294ddb859d3817afa1946c0753d43f5cb0f75b/root/etc/cont-init.d/50-certbot#L192-L219

The section for revoking might need some edits as well.

tashian replied:
I see. Ok, I've made these edits.

As far as I can tell, there's no EAB credential state stored in the `ORIG` variables, so I can't do EAB revocation using a custom ACME server. The way it works for ZeroSSL is via a custom lookup API that maps the email to an EAB account. But, that seems to be a ZeroSSL thing, not part of the ACME standard.

So, for now, maybe EAB should not be allowed for custom ACME servers?

tashian commented on root/etc/cont-init.d/50-certbot:
```diff
@@ -190,14 +213,20 @@ rm -rf /config/keys/cert.key
-    if [ "$ORIGCERTPROVIDER" = "zerossl" ] && [ -n "$ORIGEMAIL" ]; then
+    if [ "$ORIGCERTPROVIDER" = "custom" ]; then
+		REV_ACMESERVER=$ORIGACMEURL
+		if [ -n "$ORIGACMECABUNDLE" ]; then
+			echo "$ORIGACMECABUNDLE" | base64 -d - > /config/origcabundle.pem
+			export REQUESTS_CA_BUNDLE="/config/origcabundle.pem"
```
Hi @aptalca,

For revocation, I've exported the original CA bundle to `/config/origcabundle.pem` here and I've set `REQUESTS_CA_BUNDLE` so that it's picked up by `certbot`/`curl`. 

Will I need to re-set `REQUESTS_CA_BUNDLE` after revocation, so it points to the contents of `/config/cabundle.pem` again?
